respy/simulate.py

```diff
@@ -205,33 +207,83 @@ def simulate(
 
         wages = state_space.get_attribute_from_period("wages", period)
         nonpecs = state_space.get_attribute_from_period("nonpecs", period)
-        index_to_choice_set = state_space.get_attribute_from_period(
-            "dense_key_to_choice_set", period
+        index_to_complex = state_space.get_attribute_from_period(
+            "dense_key_to_complex", period
         )
         continuation_values = state_space.get_continuation_values(period=period)
 
         current_df_extended = _simulate_single_period(
             current_df,
-            index_to_choice_set,
+            index_to_complex,
             wages,
             nonpecs,
             continuation_values,
             optim_paras=optim_paras,
+            options=options,
         )
 
-        data.append(current_df_extended.copy(deep=True))
+        if optim_paras["exogenous_processes"]:
+            data.append(
+                current_df_extended.copy(deep=True).drop(
+                    "dense_key_next_period", axis=1
+                )
+            )
+        else:
+            data.append(current_df_extended.copy(deep=True))
 
         if is_n_step_ahead and period != n_simulation_periods - 1:
             current_df_extended = current_df_extended.reset_index()
             df = apply_law_of_motion_for_core(current_df_extended, optim_paras)
             state_space_columns = create_state_space_columns(optim_paras)
+
+            if optim_paras["exogenous_processes"]:
+                dense_state_columns = create_dense_state_space_columns(optim_paras)
+                df = update_dense_state_variables(
+                    df,
+                    state_space.dense_key_to_dense_covariates,
+                    optim_paras["exogenous_processes"].keys(),
+                    dense_state_columns,
+                )
             df = df.set_index(["identifier", "period"])[state_space_columns]
 
     simulated_data = _process_simulation_output(data, optim_paras)
 
     return simulated_data
 
 
+def update_dense_state_variables(
+    df, dense_key_to_dense_covariates, exogenous_processes, dense_states
+):
+    """Update the value of the exogenous processes.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        A pandas DataFrame containing the updated state variables, as well as the
+        draw of next periods dense key.
+    dense_key_to_dense_covariates : dict
+        Dictionary with dense_key as keys and dense grid point as values.
+    exogenous_processes: list
+        List of all exogenous processes.
+    dense_states : list
+        List of all dense variables.
+
+    Returns
+    -------
+    df : pandas.DataFrame
+        A pandas DataFrame containing the updated state variables and the updated
+        exogenous process.
+    """
+    for dense_key in df["dense_key_next_period"].unique():
+        for dense_pos, dense_variable in enumerate(dense_states):
+            if dense_variable in exogenous_processes:
```

Comment:
Maybe we can get rid of this if statement by directly starting with the exogenous processes instead of first looping for the static dense vars. We could say something like:
for dense_pos, dense_key in enumerate(dense_states(len(optim_paras["exogenous_processes"])):
...

Reply:
But I need the exact position of the dense variable to select from the dense covariates!

Reply:
Then just add the number of static dense vars to dense_pos?